src/Microsoft.Framework.Configuration/ConfigurationBase.cs

```diff
@@ -23,7 +23,7 @@ public ConfigurationBase(IList<IConfigurationProvider> providers)
 
         public abstract string Path { get; }
 
-        public string this[string key]
+        public virtual string this[string key]
```

Comment:
I don't have any fundamental reason to push back against making things virtual, but I believe the scenario you describe in the description of the PR and also what you have in the test are better addressed creating your own custom IConfigurationProvider and registering with the builder. That way not only the indexer will return the custom value butl also ConfigurationSection.Value, and also your source will be able to accept new values from the setters.